hpx/lcos/channel.hpp

```diff
@@ -0,0 +1,397 @@
+#include <utility>
+
+namespace hpx { namespace lcos
+{
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename T = void> class channel;
+    template <typename T = void> class receive_channel;
+    template <typename T = void> class send_channel;
+
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename T, typename Channel>
+    class channel_iterator
+      : public hpx::util::iterator_facade<
+            channel_iterator<T, Channel>, T const, std::input_iterator_tag>
+    {
+        typedef hpx::util::iterator_facade<
+                channel_iterator<T, Channel>, T const, std::input_iterator_tag
+            > base_type;
+
+    public:
+        channel_iterator()
+          : channel_(nullptr), data_(T(), false)
+        {}
+
+        channel_iterator(Channel const& c)
+          : channel_(&c), data_(get_checked())
+        {}
+
+    private:
+        std::pair<T, bool> get_checked() const
+        {
+            hpx::future<T> f = channel_->get_async();
+            f.wait();
+            if (!f.has_exception())
+            {
+                return std::make_pair(f.get(), true);
+            }
+            return std::make_pair(T(), false);
+        }
+
+        friend class hpx::util::iterator_core_access;
+
+        bool equal(channel_iterator const& rhs) const
+        {
+            return (data_.second == rhs.data_.second &&
+                        (channel_ == rhs.channel_ ||
+                            (channel_ != nullptr && rhs.channel_ != nullptr &&
+                             channel_->get_id() == rhs.channel_->get_id())
+                        )
+                    ) ||
+                (!data_.second && rhs.channel_ == nullptr) ||
+                (channel_ == nullptr && !rhs.data_.second);
+        }
+
+        void increment()
+        {
+            if (channel_)
+                data_ = get_checked();
+        }
+
+        typename base_type::reference dereference() const
+        {
+            HPX_ASSERT(data_.second);
+            return data_.first;
+        }
+
+    private:
+        Channel const* channel_;
+        std::pair<T, bool> data_;
+    };
+
+    template <typename Channel>
+    class channel_iterator<void, Channel>
+      : public hpx::util::iterator_facade<
+            channel_iterator<void, Channel>, util::unused_type const,
+            std::input_iterator_tag>
+    {
+        typedef hpx::util::iterator_facade<
+                channel_iterator<void, Channel>, util::unused_type const,
+                std::input_iterator_tag
+            > base_type;
+
+    public:
+        channel_iterator()
+          : channel_(nullptr), data_(false)
+        {}
+
+        channel_iterator(Channel const& c)
+          : channel_(&c), data_(get_checked())
+        {}
+
+    private:
+        bool get_checked()
+        {
+            hpx::future<void> f = channel_->get_async();
+            f.wait();
```

Comment:
True, without `co_await`, this is hard to implement. Why not have `future<T>` as value type directly? This won't prohibit the use of `co_await` in future code, and it's the user's responsibility to deal with the future: block immediately, or create some kind of awkward continuation based looping infrastructure. I am wondering if we can't offer a high level solution to loop over future values though....